### ISSUE TITLE
Dataset dates for remote files gets stale

### ISSUE THREAD

**Issue by cjhendrix** [OCHA-DAP/hdx-ckan#2804]
The resource shown below is being updated daily, but our date for it will stay the same, which is confusing. 

We could:
1. figure out a way to monitor remote resources to see if they have changed, though that may require different tests for different remote sources
2. have a way for contributors to mark a dataset as "frequently updated" in which case we don't show a date.  Although easier, this would be somewhat confusing to users and could also get stale if data updates stop.

**Comment by danmihaila** [comment]
@cjhendrix pls give more details about what resource is this issue about. Do we have this information in activity stream displayed for dataset owning the resource?

**Comment by pjdufour** [comment]
This applies to remotely hosted OSM extracts which are being updated every 30 minutes, such as:

https://data.hdx.rwlabs.org/dataset/nepal-openstreetmap-extracts-roads
https://data.hdx.rwlabs.org/dataset/nepal-openstreetmap-extracts-buildings
https://data.hdx.rwlabs.org/dataset/nepal-openstreetmap-extracts-places

I currently go in and manually update the date each day.  Don't know best way to deal with this.  Downloading a 200MB+ file to hash and compare to existing hash seems easy to code but inefficient.

With OGC services (such as GeoNode), you could sync with a dataset with a capabilities doc or CSW endpoint at an interval and specify which fields to update.  However, for datasets without corresponding capabilities docs, it's not as straight forward.

There's no structured field for "frequency" in the UI, so that might be a first step.  I currently add the frequency to the description.

**Comment by cjhendrix** [comment]
ST asked for this to be done "soon" in our discussion on the FAQ text.  Putting it into sprint candidates.  Tagging @amcguire62.

**Comment by davidmegginson** [comment]
We can handle freshness with a separate monitoring system that works like this:

![Update sequence diagram](http://www.websequencediagrams.com/cgi-bin/cdraw?lz=cGFydGljaXBhbnQgIkV4dGVybmFsIHNpdGVzIiBhcyBTb3VyY2VzCgAbDU1vbml0b3Jpbmcgc3lzdGVtACYFAAwKACAOSERYIEFQSQBKBUhEWAoKbG9vcCBmb3IgZWFjaCBkYXRhc2V0CiAANgstPkhEWDogcmVxdWVzdAAYC2FjdGl2YXRlAGEMAAsLSERYCiAgSERYLT4AgRoKOiBkb3dubG9hZABkCCByZWNvcmQKICBkZQAtDwB4DAA3DGNoZWNrIHVwZGF0ZSBmAIEUBW5jeQogIG9wdCB0aW1lAIFMBWZyZXNobmVzcwAqBgogIACBXg5yZXMAgksFCiAgADQGAAkIIGhhcyBhbiBlAIJ6CFVSTAAgBQCCCA4AgwcHAIIQCgBEDQCCFAsAgy8IAD0GAIM-BwCBfhcANQ8AgggLADgOAIIDGHNhdmUgdG8gZGlzawAMH2NvbXBhcgApBWxhc3QAgxAJAIFkB29wdCBkaWZmZXJzIGZyb20AHwZjb3B5AIIFBwCEFw0AhB4HAIMNBwCDUwhkYXQAggYIZW4AUwYAAgZlbmQKZW5kCg&s=modern-blue)

This should definitely run on a separate VM.  Over the week (probably with one batch nightly), the monitoring system would download each dataset metadata record once and compare its update frequency with the last date it was checked; if a check is due, then we download any resources that have external URLs and compare them (simple Unix cmp(1) command) with the last version we downloaded.  If it's changed, update the date.

**Comment by davidmegginson** [comment]
@brew mentions the CKAN [deadoralive](https://github.com/ckan/ckanext-deadoralive) extension as another possible model to follow.

**Comment by reubano** [comment]
Took a quick look at the deadoralive source and, despite what it says, it looks pretty [synchronus](https://github.com/ckan/deadoralive/blob/master/deadoralive/deadoralive.py#L193). Could be an issue at some point.

**Comment by davidmegginson** [comment]
Thanks for checking that, @reubano — I've seen similar things in Drupal, etc, so I can imagine how it's set up.

In any case, as a basic architectural principal, we should never turn down a chance for horizontal scaling when the cost is this low; there's no reason to run these housekeeping batch tasks inside the main CKAN process, since they're not part of the interactive user experience, while there are strong reasons for running batch jobs externally:
1. **Performance:** not competing for the same CPU/memory/diskspace with the Web layer (except for those required to respond to API calls).
2. **Stability:** a bug in the housekeeping/batch code won't cause a crash or security hole in the HDX web site.
3. **Maintainability:** releases to the batch housekeeping code don't have to be tightly coupled to releases of the Web site.

**Comment by reubano** [comment]
@davidmegginson agreed. deadoralive is def better than the current situation. A few options for future consideration are [uptime](https://github.com/fzaninotto/uptime) or, if it must be python, [twisted](http://as.ynchrono.us/2006/05/limiting-parallelism_22.html)/[tornado](http://stackoverflow.com/a/25549675/408556)/[grequests](https://github.com/kennethreitz/grequests).
